contracts/crowdsale/price/CapStagedCrowdsale.sol

```diff
@@ -0,0 +1,93 @@
+pragma solidity ^ 0.4.24;
+
+import "../Crowdsale.sol";
+import "../../math/SafeMath.sol";
+import "../../ownership/Ownable.sol";
+import "../validation/CappedCrowdsale.sol";
+
+
+/**
+  * @title CapStagedCrowdsale
+  * @dev Extension of Crowdsale contract that changes the price of tokens regarding of raised ETH. Each stage of crowdsale is defined in StageLimit (cap limit for the stage) and stageRate (stage rate)
+  * Stages must be inserted from first to last with increasing stageLimit.
+  */
+contract CapStagedCrowdsale is CappedCrowdsale {
+  using SafeMath for uint256;
+
+  struct Stage {
+    uint256 limit;
+    uint256 rate;
+  }
+
+  Stage[] public stages;
+
+  /**
+   * @param _initialRate Number of tokens a buyer gets per wei - default parameter for crowdsale contract
+   * @param _wallet Address where funds should be transferred
+   * @param _token ERC20 token address
+   * @param _cap max cap that can be reached in wei
+   * @param _stageLimits Array of stage limits in wei
+   * @param _stageRates Array of rates for every stage
+   */
+  constructor
+  (
+    uint256 _initialRate,
+    address _wallet,
+    ERC20 _token,
+    uint256 _cap,
+    uint256[] _stageLimits,
+    uint256[] _stageRates
+  )
+    CappedCrowdsale(_cap)
+    Crowdsale(_initialRate, _wallet, _token)
+    public
+  {
+    require(_stageLimits.length == _stageRates.length);
+    require(_stageLimits.length > 0);
+    for (uint256 i = 0; i < _stageLimits.length; i++) {
+      if (i+1 < _stageLimits.length) {
+        if (_stageLimits[i] < _stageLimits[i+1]) {
+          stages.push(Stage({
+            limit: _stageLimits[i],
+            rate: _stageRates[i]
+          }));
+        }
+      } else {
+        if (_stageLimits[i] > _stageLimits[i-1]) {
+          stages.push(Stage({
+            limit: _stageLimits[i],
+            rate: _stageRates[i]
+          }));
+        } else {
+          revert();
```

Comment:
I find this confusing. It's not clear what conditions will trigger a revert.
I would prefer to do all the validations first. Have a statement that checks that the limits array is ordered by increasing value.